File: shimbel.cpp
#include "shimbel.h"

#include <algorithm>
#include <climits>

// Sentinels chosen to avoid overflow on addition
static const int POS_INF = INT_MAX / 2;
static const int NEG_INF = INT_MIN / 2;

// Min-plus matrix multiplication:  C[i][j] = min_m ( A[i][m] + B[m][j] )
static std::vector<std::vector<int>> minPlusMul(
    const std::vector<std::vector<int>>& A,
    const std::vector<std::vector<int>>& B, int n)
{
    std::vector<std::vector<int>> C(n, std::vector<int>(n, POS_INF));
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j)
            for (int m = 0; m < n; ++m)
                if (A[i][m] < POS_INF && B[m][j] < POS_INF)
                    C[i][j] = std::min(C[i][j], A[i][m] + B[m][j]);
    return C;
}

// Max-plus matrix multiplication:  C[i][j] = max_m ( A[i][m] + B[m][j] )
static std::vector<std::vector<int>> maxPlusMul(
    const std::vector<std::vector<int>>& A,
    const std::vector<std::vector<int>>& B, int n)
{
    std::vector<std::vector<int>> C(n, std::vector<int>(n, NEG_INF));
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j)
            for (int m = 0; m < n; ++m)
                if (A[i][m] > NEG_INF && B[m][j] > NEG_INF)
                    C[i][j] = std::max(C[i][j], A[i][m] + B[m][j]);
    return C;
}

std::vector<std::vector<int>> Shimbel::minPath(
    const std::vector<std::vector<int>>& weightMatrix, int k)
{
    const int n = static_cast<int>(weightMatrix.size());
    if (n == 0 || k <= 0)
        return std::vector<std::vector<int>>(n, std::vector<int>(n, POS_INF));

    // Build initial weight matrix W (POS_INF = no edge)
    std::vector<std::vector<int>> W(n, std::vector<int>(n, POS_INF));
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j)
            if (weightMatrix[i][j] != 0)
                W[i][j] = weightMatrix[i][j];

    auto result = W;  // S(1) = W
    for (int step = 2; step <= k; ++step)
        result = minPlusMul(result, W, n);

    return result;
}

std::vector<std::vector<int>> Shimbel::maxPath(
    const std::vector<std::vector<int>>& weightMatrix, int k)
{
    const int n = static_cast<int>(weightMatrix.size());
    if (n == 0 || k <= 0)
        return std::vector<std::vector<int>>(n, std::vector<int>(n, NEG_INF));

    // Build initial weight matrix W (NEG_INF = no edge)
    std::vector<std::vector<int>> W(n, std::vector<int>(n, NEG_INF));
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j)
            if (weightMatrix[i][j] != 0)
                W[i][j] = weightMatrix[i][j];

    auto result = W;  // S(1) = W
    for (int step = 2; step <= k; ++step)
        result = maxPlusMul(result, W, n);

    return result;
}

File: web/js/shimbel.js
window.Shimbel = (function () {
  var POS_INF = 1e9;
  var NEG_INF = -1e9;

  function minPlusMul(A, B, n) {
    var C = [];
    for (var i = 0; i < n; i++) {
      C.push(new Array(n).fill(POS_INF));
    }
    for (var i = 0; i < n; i++) {
      for (var j = 0; j < n; j++) {
        for (var m = 0; m < n; m++) {
          if (A[i][m] !== POS_INF && B[m][j] !== POS_INF) {
            var s = A[i][m] + B[m][j];
            if (s < C[i][j]) {
              C[i][j] = s;
            }
          }
        }
      }
    }
    return C;
  }

  function maxPlusMul(A, B, n) {
    var C = [];
    for (var i = 0; i < n; i++) {
      C.push(new Array(n).fill(NEG_INF));
    }
    for (var i = 0; i < n; i++) {
      for (var j = 0; j < n; j++) {
        for (var m = 0; m < n; m++) {
          if (A[i][m] !== NEG_INF && B[m][j] !== NEG_INF) {
            var s = A[i][m] + B[m][j];
            if (s > C[i][j]) {
              C[i][j] = s;
            }
          }
        }
      }
    }
    return C;
  }

  function minPath(weightMatrix, k) {
    var n = weightMatrix.length;
    // Build initial matrix W: 0 entries become POS_INF, others kept
    var W = [];
    for (var i = 0; i < n; i++) {
      W.push(new Array(n));
      for (var j = 0; j < n; j++) {
        W[i][j] = (weightMatrix[i][j] === 0) ? POS_INF : weightMatrix[i][j];
      }
    }

    // Multiply k-1 times: result = W^k in min-plus semiring
    var result = W;
    for (var step = 1; step < k; step++) {
      result = minPlusMul(result, W, n);
    }
    return result;
  }

  function maxPath(weightMatrix, k) {
    var n = weightMatrix.length;
    // Build initial matrix W: 0 entries become NEG_INF, others kept
    var W = [];
    for (var i = 0; i < n; i++) {
      W.push(new Array(n));
      for (var j = 0; j < n; j++) {
        W[i][j] = (weightMatrix[i][j] === 0) ? NEG_INF : weightMatrix[i][j];
      }
    }

    // Multiply k-1 times: result = W^k in max-plus semiring
    var result = W;
    for (var step = 1; step < k; step++) {
      result = maxPlusMul(result, W, n);
    }
    return result;
  }

  return {
    POS_INF: POS_INF,
    NEG_INF: NEG_INF,
    minPlusMul: minPlusMul,
    maxPlusMul: maxPlusMul,
    minPath: minPath,
    maxPath: maxPath
  };
})();
